<%
      import djerba.plugins.pancurx.constants as plugin
      from djerba.plugins.pancurx.html import make_somatic_table_rows, k_comma_format, make_signature_table_rows
      from djerba.util.html import html_builder
      loads = results.get('loads')
      sigs = results.get('sigs')
      reportable_variants = results.get('reportable_variants')
      all_variants = results.get('all_variants')
      ploidy = results.get('ploidy')
      tier_mode = results.get('tier_mode')

%>


<h2 class='lbr'>Somatic Genome</h2>
<!-- <div style="padding:0px 0px 0px 50px; break-inside: avoid;display: table;"> -->

<div style="padding:0px 0px 0px 50px; ">
<h3 class="lbr">Mutational Status of Common Biliary Driver and Targetable Genes</h3>

      <p class="lbr">Below are listed the mutational status of genes that are commonly altered in biliary tract cancers. 
      Actionable alterations are listed regardless of their mutational status, while most oncogenes are only listed if mutated. 
      * donates genes of the discovery tier, which are only included when less than 7 confident hits are found.</p>	

      <table class="lbr sortable" style="width:100%" >
      <tr><td style="width:8%">Gene</td><td style="width:5%">Chr.</td><td style="width:5%">CN</td><td style="width:5%">A|B</td><td style="width:5%">tVAF</td><td style="width:62%">Variant</td></tr>

            % for row in make_somatic_table_rows(reportable_variants, ploidy, tier_mode):
                  ${row}
            % endfor

      </table>

</div>
      
       
<div style="padding:0px 0px 0px 50px; ">
<h3 class="lbr">Mutational Load</h3>

      <p class="lbr">In this tumour, the <b>TMB (genome-wide) is ${loads.get('TMB')} mutations / mb</b>.  <br><br>
      Tumour Mutation Burden (TMB) represents the number of non-inherited mutations per million bases (Mb) of genomic sequence in tumor tissue. 
      Essentially, it quantifies how many mutations occur in the genome of cancer cells. TMB has gained prominence due to its association with immunotherapy response.
      Below are the comparisons of this sample with the LBR cohort for each mutation type:<p>

      <table class="lbr" style="width:100%">
            <tr>
                  <td style="text-align:center"><b>${k_comma_format(loads.get('snv_count'))} SNVs (${loads.get('snv_percentile')}% PR)</b></td>
                  <td style="text-align:center"><b>${k_comma_format(loads.get('indel_count'))} In/Dels (${loads.get('indel_percentile')}% PR)</b></td>
                  <td style="text-align:center"><b>${k_comma_format(loads.get('sv_count'))} SVs (${loads.get('sv_percentile')}% PR)</b></td>
            </tr>
            <tr>
                  <td style="text-align:center"><img id='HISTBOX_SNV' style='width: 100%; object-fit: contain' src="${results.get(plugin.HISTBOX_SNV)}"/></td>
                  <td style="text-align:center"><img id='HISTBOX_INDEL' style='width: 100%; object-fit: contain' src="${results.get(plugin.HISTBOX_INDEL)}"/></td>
                  <td style="text-align:center"><img id='HISTBOX_SV' style='width: 100%; object-fit: contain' src="${results.get(plugin.HISTBOX_SV)}"/></td>
            </tr>
      </table>

</div>

<div style="padding:0px 0px 0px 50px; ">
      <h3 class="lbr">Mutational Signatures</h3>
      <p class="lbr">Mutational signatures can be used as a physiological readout of the biological history of a cancer 
            and also have potential use for discerning ongoing mutational processes from historical ones, thus possibly revealing new targets for anticancer therapies.
            Signatures are a decomposition of mutations by their flanking bases (ie the base on either side of the mutation) and the change itself (eg. C to A or T to G).
            The plot on the left shows the proportion (%) of each mutation change (eg. C>A in red) and within that are further subset the flanking bases (A_C). 
          On the right are listed the proportion of single base pair substitutions assigned to COSMIC (version 3) signatures and that signature's aetiologies.</p>
          
          
      <div style=" break-inside: avoid;display: table;"> 


                  <img id='SNV_CONTEXT_PLOT' style='height: 300px; position: relative; float:left; margin-left: 50px; margin-right: 50px; ' src="${results.get(plugin.SNV_CONTEXT_PLOT)}"/>


                  <table class="lbr sortable">
                        <tr><th style="width: 100px"></th><th style="width: 150px">Proposed Aetiology</th><th style="width: 100px">Proportion</th>
                        % for row in make_signature_table_rows(sigs):
                              ${row}
                        % endfor
            
            
                  </table>
                  <br><br>
      </div>



</div>

