<%
      import djerba.plugins.pancurx.constants as plugin
      from djerba.plugins.pancurx.html import make_somatic_table_rows, make_signature_table_rows
      from djerba.util.html import html_builder
      loads = results.get('loads')
      sigs = results.get('sigs')
      reportable_variants = results.get('reportable_variants')
      all_variants = results.get('all_variants')
      ploidy = results.get('ploidy')
      tier_mode = results.get('tier_mode')
      sigs = results.get('sigs')
%>

<div id="main">

<h2 class="pcx">Somatic Mutations</h2>

<h3 class="pcx">Whole Genome Summary</h3>

      <img id='DSBR' style='width: 100%; object-fit: contain' src="${results.get(plugin.WHOLE_GENOME_PLOT)}"/>
      <p style="page-break-after:always;"></p>


<h3 class="pcx">Mutational Load</h3>

<table style="width:100%">
	<table>
		<tr>
		<td colspan="3"><b>TMB (genome-wide):</b> ${loads.get('TMB')} mutations / mb</td>
		</tr>
		<tr>
		<td style="width:33%"><b>Nonsynonymous SNVs:</b> ${loads.get('nonsyn_count')} </td>
		<td style="width:33%"><b>Frameshift deletions:</b> ${loads.get('del_frameshift_count')}  </td>
		<td style="width:33%"><b>Genes with Deletion BPs:</b> ${loads.get('sv_del_bp_gene_count')}  </td>
		</tr>
		<tr>
		<td><b>Stopgain SNVs:</b> ${loads.get('stopgain_count')}  </td>
		<td><b>Nonframeshift deletions:</b> ${loads.get('del_nonframeshift_count')}  </td>
		<td><b>Genes with Duplication BPs:</b> ${loads.get('sv_dup_bp_gene_count')}  </td>
		</tr>
		<tr>
		<td><b>Stoploss SNVs:</b> ${loads.get('stoploss_count')} </td>
		<td><b>Frameshift Insertions:</b> ${loads.get('ins_frameshift_count')} </td>
		<td><b>Genes with Inversion BPs:</b> ${loads.get('sv_inv_bp_gene_count')} </td>
		</tr>
		<tr>
		<td><b>Splicing SNVs:</b> ${loads.get('splice_count')}  </td>
		<td><b>Nonframeshift Insertions:</b> ${loads.get('ins_nonframeshift_count')} </td>
		<td><b>Genes with Translocation BPs:</b> ${loads.get('sv_tra_bp_gene_count')} </td>
		</tr>
		<tr>
		<td><b>Total Deleterious SNVs:</b> ${loads.get('total_del_snv')}</td>
		<td><b>Total Deleterious Indels:</b> ${loads.get('total_del_indel')}</td>
		<td><b>Total Deleterious SV Breakpoints:</b> ${loads.get('total_del_sv')}</td>
		</tr>
		</table>
		
		<br>
		
      <table style="width:100%"><tr>
      <tr>
      <td style="text-align:center"><b>${loads.get('snv_count')} SNVs (${loads.get('snv_percentile')}% PR)</b></td>
      <td style="text-align:center"><b>${loads.get('indel_count')} In/Dels (${loads.get('indel_percentile')}% PR)</b></td>
      <td style="text-align:center"><b>${loads.get('sv_count')} SVs (${loads.get('sv_percentile')}% PR)</b></td>
      </tr>

      <td style="text-align:center"><img id='HISTBOX_SNV' style='width: 100%; object-fit: contain' src="${results.get(plugin.HISTBOX_SNV)}"/></td>
      <td style="text-align:center"><img id='HISTBOX_INDEL' style='width: 100%; object-fit: contain' src="${results.get(plugin.HISTBOX_INDEL)}"/></td>
      <td style="text-align:center"><img id='HISTBOX_SV' style='width: 100%; object-fit: contain' src="${results.get(plugin.HISTBOX_SV)}"/></td>

      </tr></table>

	<table style="width:100%"><tr>

            <td style="text-align:center"><img id='INDEL_BIN_PLOT' style='width: 40%; object-fit: contain' src="${results.get(plugin.INDEL_BIN_PLOT)}"/></td>
		<td style="text-align:center"><img id='SV_BIN_PLOT' style='width: 40%; object-fit: contain' src="${results.get(plugin.SV_BIN_PLOT)}"/></td>
  
	</tr></table>



      <img id='SNV_VAF_PLOT' style='width: 100%; object-fit: contain' src="${results.get(plugin.SNV_VAF_PLOT)}"/>
      <img id='INDEL_VAF_PLOT' style='width: 100%; object-fit: contain' src="${results.get(plugin.INDEL_VAF_PLOT)}"/>

	  <h3 class="pcx">Mutational Signatures</h3>
          
		<img id='SNV_CONTEXT_PLOT' style='height: 300px; position: relative; float:left; margin-left: 50px; margin-right: 50px; ' src="${results.get(plugin.SNV_CONTEXT_PLOT)}"/>


		<table class="lbr sortable">
			<tr><th style="width: 100px"></th><th style="width: 150px">Proposed Aetiology</th><th style="width: 100px">Proportion</th>
			% for row in make_signature_table_rows(sigs):
					${row}
			% endfor


		</table>
		<br>
		<br>






      <h3 class="pcx">Driver Gene Status</h3>
	
      <table class="pcx sortable"  style="width:100%" >
		<tr><td><b>Gene</b></td><td><b>Chr.</b></td><td><b>Copy Number</b></td><td><b>A|B Count</b></td><td><b>tVAF</b></td><td><b>Variant</b></td></tr>

            % for row in make_somatic_table_rows(reportable_variants, ploidy, tier_mode):
                  ${row}
            % endfor

      </table>
      <p>CN values in orange show copy gains defined as ploidy + 1, in red amplifications as ploidy x 4, in light blue CN-LOH as CN less than 1.5, and in dark blue homozygous deletions as CN less than 0.5. Copy Numbers and A.B ratios annotated with a * represent instances where we estimate more than one copy state within the gene.</p>


      
  
<p style="page-break-after:always;"></p>

<h3 class="pcx">Ploidy and Chromothripsis</h3>

      <img id='CELLULOID' style='width: 100%; object-fit: contain' src="${results.get(plugin.CELLULOID_PLOT)}"/>



</div>


