<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>djerba.metrics API documentation</title>
<meta name="description" content="Classes to contain genetic_alteration metric functions; see prototypes directory" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>djerba.metrics</code></h1>
</header>
<section id="section-intro">
<p>Classes to contain genetic_alteration metric functions; see prototypes directory</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;Classes to contain genetic_alteration metric functions; see prototypes directory&#34;&#34;&#34;

import gzip
import json
import logging
import pandas as pd
import re

from statsmodels.distributions.empirical_distribution import ECDF
from math import isnan
from djerba.utilities.base import base

class mutation_extended_gene_metrics(base):

    &#34;&#34;&#34;
    Gene-level fields for the MUTATION_EXTENDED alteration type

    Includes fields for Elba output:
    - Allele Fraction Percentage
    - Chromosome
    - Gene
    - FDA Approved Treatment
    - OncoKB
    - Protein Change
    - Variant Reads And Total Reads
    &#34;&#34;&#34;

    # MAF column headers
    HUGO_SYMBOL = &#39;Hugo_Symbol&#39;
    CHROMOSOME = &#39;Chromosome&#39;
    HGVSP_SHORT = &#39;HGVSp_Short&#39;
    T_DEPTH = &#39;t_depth&#39;
    T_ALT_COUNT = &#39;t_alt_count&#39;
    HIGHEST_LEVEL = &#39;Highest_level&#39;
    FDA_APPROVED_TREATMENT = &#39;FDA_Approved_Treatment&#39;
    LEVEL_1  = &#39;LEVEL_1&#39;
    LEVEL_2A = &#39;LEVEL_2A&#39;
    LEVEL_2B = &#39;LEVEL_2B&#39;
    LEVEL_3A = &#39;LEVEL_3A&#39;
    LEVEL_3B = &#39;LEVEL_3B&#39;
    LEVEL_4  = &#39;LEVEL_4&#39;
    LEVEL_R1 = &#39;LEVEL_R1&#39;
    LEVEL_R2 = &#39;LEVEL_R2&#39;
    REQUIRED_MAF_COLS = [
        HUGO_SYMBOL,
        CHROMOSOME,
        HGVSP_SHORT,
        T_DEPTH,
        T_ALT_COUNT,
    ]
    BIOMARKER_LEVELS = [
        LEVEL_1,
        LEVEL_2A,
        LEVEL_2B,
        LEVEL_3A,
        LEVEL_3B,
        LEVEL_4,
        LEVEL_R1,
        LEVEL_R2
    ]

    # Output headers
    PROTEIN_CHANGE = &#39;Protein_Change&#39;
    ALLELE_FRACTION_PERCENTILE = &#39;Allele_Fraction_Percentile&#39;
    VARIANT_READS_AND_TOTAL_READS = &#39;Variant_Reads_And_Total_Reads&#39;
    ONCOKB = &#39;OncoKB&#39;

    def __init__(self, maf_path, log_level=logging.WARNING, log_path=None):
        self.logger = self.get_logger(log_level, &#34;%s.%s&#34; % (__name__, type(self).__name__), log_path)
        cols = self._find_columns_to_use(maf_path)
        df = pd.read_csv(maf_path, usecols=cols, delimiter=&#34;\t&#34;, comment=&#34;#&#34;)
        if df.size == 0:
            msg = &#34;No rows found in MAF input %s&#34; % maf_path
            self.logger.error(msg)
            raise RuntimeError(msg)
        else:
            self.logger.debug(&#34;Found %d MAF rows from input path %s&#34; % (df.size, maf_path))
        self.metrics = {}
        for index, row in df.iterrows():
            total = row[self.T_DEPTH]
            variants = row[self.T_ALT_COUNT]
            try:
                afp = (float(variants)/float(total))*100
            except ZeroDivisionError:
                afp = &#34;NA&#34;
            oncokb = row.get(self.HIGHEST_LEVEL)
            if oncokb == None:
                oncokb = &#34;NA&#34;
            self.metrics[row[self.HUGO_SYMBOL]] = {
                self.CHROMOSOME: row[self.CHROMOSOME],
                self.PROTEIN_CHANGE: row[self.HGVSP_SHORT],
                self.VARIANT_READS_AND_TOTAL_READS: &#34;%d/%d&#34; % (variants, total),
                self.ALLELE_FRACTION_PERCENTILE: afp,
                self.ONCOKB: oncokb,
                self.FDA_APPROVED_TREATMENT: self._find_treatment_string(row)
            }
        metric_string = json.dumps(self.metrics[row[self.HUGO_SYMBOL]], sort_keys=True)
        self.logger.debug(&#34;Example of metrics for gene %s: %s&#34; % (row[self.HUGO_SYMBOL], metric_string))
        self.logger.info(&#34;Found mutation_extended metrics for %i genes&#34; % len(self.metrics))

    def _find_columns_to_use(self, maf_path):
        &#34;&#34;&#34;
        Read column headers from the MAF file, to determine what columns to input in pandas
        Handles the case where not all OncoKB LEVEL columns are present
        MAF files may have 100+ columns, so this is preferred to reading in the whole file
        &#34;&#34;&#34;
        if re.search(&#34;\.gz$&#34;, maf_path):
            in_file = gzip.open(maf_path, &#39;rt&#39;)
        else:
            in_file = open(maf_path, &#39;r&#39;)
        header_line = None
        # read the first non-comment line to get column headers
        while True:
            line = in_file.readline()
            if line == &#39;&#39;: # end of file
                break
            elif re.match(&#34;#&#34;, line):
                continue
            else:
                header_line = line
                break
        in_file.close()
        if not header_line:
            msg = &#34;No header line found in MAF file &#39;%s&#39;&#34; % maf_path
            self.logger.error(msg)
            raise RuntimeError(msg)
        column_names = re.split(&#34;\t&#34;, line.strip())
        self.logger.debug(&#34;Found MAF column names: &#34;+&#34;, &#34;.join(column_names))
        missing = []
        for name in self.REQUIRED_MAF_COLS:
            if not name in column_names:
                missing.append(name)
        if len(missing) &gt; 0:
            msg = &#34;Missing required MAF columns: &#34;+&#34;, &#34;.join(missing)
            self.logger.error(msg)
            raise RuntimeError(msg)
        use_cols = self.REQUIRED_MAF_COLS.copy() # list of columns to read
        optional_total = 0
        optional_cols = [self.HIGHEST_LEVEL]
        optional_cols.extend(self.BIOMARKER_LEVELS)
        for name in optional_cols:
            if name in column_names:
                use_cols.append(name)
                optional_total += 1
        if optional_total == 0:
            msg = &#34;No optional MAF columns found. OncoKB annotation not applied? &#34;+\
                  &#34;OncoKB output fields will receive &#39;NA&#39; values.&#34;
            self.logger.warning(msg)
        self.logger.debug(&#34;Found MAF columns for input: &#34;+&#34;, &#34;.join(use_cols))
        return use_cols

    def _find_treatment_string(self, row):
        &#34;&#34;&#34;
        Find the &#39;FDA Approved Treatment&#39; string
        Concatenation of level codes and approved treatments
        &#34;&#34;&#34;
        # This string is a temporary placeholder, will be replaced by a more complex structure
        # See https://jira.oicr.on.ca/browse/GCGI-69
        biomarker_values = []
        for level in self.BIOMARKER_LEVELS:
            value = row.get(level) # will be None if no column in MAF file for this level
            if not self.is_null(value):
                biomarker_values.append(&#39;%s:%s&#39; % (level, value))
        if len(biomarker_values) == 0:
            return &#39;NA&#39;
        else:
            return &#39;;&#39;.join(biomarker_values)

    def get_metrics_for_gene(self, gene):
        &#34;&#34;&#34;Get metric dictionary for the named gene, or None if gene is not present&#34;&#34;&#34;
        return self.metrics.get(gene)

    def get_metrics(self):
        return self.metrics

class mutation_extended_sample_metrics(base):

    &#34;&#34;&#34;
    Sample-level metrics for the MUTATION_EXTENDED alteration type

    Includes metrics for Elba output:
    - TMB_PER_MB
    &#34;&#34;&#34;
    
    def __init__(self, maf_path, bed_path, tcga_path, cancer_type, log_level=logging.WARNING, log_path=None):
        self.logger = self.get_logger(log_level, &#34;%s.%s&#34; % (__name__, type(self).__name__), log_path)
        self.tmb = self._find_tmb(maf_path, bed_path)        
        [self.tcga_tmb, self.tcga_cohort_tmb] = self._read_tcga(tcga_path, cancer_type)
        self.tcga_pct = ECDF(self.tcga_tmb)(self.tmb)
        self.cohort_pct = ECDF(self.tcga_cohort_tmb)(self.tmb)
        
    def _find_tmb(self, maf_path, bed_path):
        &#34;&#34;&#34;Find the TMB metric: tumor mutation burden / megabase&#34;&#34;&#34;
        names = [&#39;chrom&#39;, &#39;start&#39;, &#39;end&#39;]
        maf = pd.read_csv(maf_path, sep=&#39;\t&#39;, skiprows=1)
        bed = pd.read_csv(bed_path, sep=&#39;\t&#39;, skiprows=2, header=None, names=names)
        target_space = sum(bed[&#39;end&#39;] - bed[&#39;start&#39;]) / 1000000
        keep = [
            &#39;Missense_Mutation&#39;,
            &#39;Frame_Shift_Del&#39;,
            &#39;In_Frame_Del&#39;,
            &#39;Frame_Shift_Ins&#39;,
            &#39;In_Frame_Ins&#39;,
            &#39;Splice_Site&#39;,
            &#39;Translation_Start_Site&#39;,
            &#39;Nonsense_Mutation&#39;,
            &#39;Nonstop_Mutation&#39;
        ]
        tmb = len(maf.loc[maf[&#34;Variant_Classification&#34;].isin(keep)]) / target_space
        return tmb

    def _read_tcga(self, tcga_path, cancer_type):
        &#34;&#34;&#34;Read the TCGA path and return a pair of dictionaries for TMB&#34;&#34;&#34;
        tcga_cols = [&#39;sample&#39;, &#39;tcgaTMB&#39;, &#39;callable&#39;, &#39;cancerType&#39;]
        tcga = pd.read_csv(tcga_path, sep=&#39;\t&#39;, header=0, names=tcga_cols)
        tcga_tmb =  tcga.tcgaTMB
        tcga_cohort = tcga.loc[tcga[&#34;cancerType&#34;] == cancer_type]
        tcga_cohort = tcga_cohort.rename(columns={&#39;tcgaTMB&#39;: &#39;cohortTMB&#39;})
        tcga_cohort_tmb = tcga_cohort.cohortTMB
        return (tcga_tmb, tcga_cohort_tmb)

    def get_cosmic_sigs(self):
        &#34;&#34;&#34;
        TODO: Compute the COSMIC_SIGS metric
        - Currently implemented as an R script
        - Not yet ready for production, requires more validation
        - Return &#34;NA&#34; as a placeholder for now
        &#34;&#34;&#34;
        return &#34;NA&#34;

    def get_tmb(self):
        return self.tmb

    def get_tcga_pct(self):
        return self.tcga_pct
    
    def get_cohort_pct(self):
        return self.cohort_pct
    
    def get_tcga_tmb_as_dict(self):
        &#34;&#34;&#34;Convert the pandas Series to a dictionary&#34;&#34;&#34;
        return self.tcga_tmb.to_dict()

    def get_cohort_tmb_as_dict(self):
        &#34;&#34;&#34;Convert the pandas Series to a dictionary&#34;&#34;&#34;
        return self.tcga_cohort_tmb.to_dict()</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="djerba.metrics.mutation_extended_gene_metrics"><code class="flex name class">
<span>class <span class="ident">mutation_extended_gene_metrics</span></span>
<span>(</span><span>maf_path, log_level=30, log_path=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Gene-level fields for the MUTATION_EXTENDED alteration type</p>
<p>Includes fields for Elba output:
- Allele Fraction Percentage
- Chromosome
- Gene
- FDA Approved Treatment
- OncoKB
- Protein Change
- Variant Reads And Total Reads</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class mutation_extended_gene_metrics(base):

    &#34;&#34;&#34;
    Gene-level fields for the MUTATION_EXTENDED alteration type

    Includes fields for Elba output:
    - Allele Fraction Percentage
    - Chromosome
    - Gene
    - FDA Approved Treatment
    - OncoKB
    - Protein Change
    - Variant Reads And Total Reads
    &#34;&#34;&#34;

    # MAF column headers
    HUGO_SYMBOL = &#39;Hugo_Symbol&#39;
    CHROMOSOME = &#39;Chromosome&#39;
    HGVSP_SHORT = &#39;HGVSp_Short&#39;
    T_DEPTH = &#39;t_depth&#39;
    T_ALT_COUNT = &#39;t_alt_count&#39;
    HIGHEST_LEVEL = &#39;Highest_level&#39;
    FDA_APPROVED_TREATMENT = &#39;FDA_Approved_Treatment&#39;
    LEVEL_1  = &#39;LEVEL_1&#39;
    LEVEL_2A = &#39;LEVEL_2A&#39;
    LEVEL_2B = &#39;LEVEL_2B&#39;
    LEVEL_3A = &#39;LEVEL_3A&#39;
    LEVEL_3B = &#39;LEVEL_3B&#39;
    LEVEL_4  = &#39;LEVEL_4&#39;
    LEVEL_R1 = &#39;LEVEL_R1&#39;
    LEVEL_R2 = &#39;LEVEL_R2&#39;
    REQUIRED_MAF_COLS = [
        HUGO_SYMBOL,
        CHROMOSOME,
        HGVSP_SHORT,
        T_DEPTH,
        T_ALT_COUNT,
    ]
    BIOMARKER_LEVELS = [
        LEVEL_1,
        LEVEL_2A,
        LEVEL_2B,
        LEVEL_3A,
        LEVEL_3B,
        LEVEL_4,
        LEVEL_R1,
        LEVEL_R2
    ]

    # Output headers
    PROTEIN_CHANGE = &#39;Protein_Change&#39;
    ALLELE_FRACTION_PERCENTILE = &#39;Allele_Fraction_Percentile&#39;
    VARIANT_READS_AND_TOTAL_READS = &#39;Variant_Reads_And_Total_Reads&#39;
    ONCOKB = &#39;OncoKB&#39;

    def __init__(self, maf_path, log_level=logging.WARNING, log_path=None):
        self.logger = self.get_logger(log_level, &#34;%s.%s&#34; % (__name__, type(self).__name__), log_path)
        cols = self._find_columns_to_use(maf_path)
        df = pd.read_csv(maf_path, usecols=cols, delimiter=&#34;\t&#34;, comment=&#34;#&#34;)
        if df.size == 0:
            msg = &#34;No rows found in MAF input %s&#34; % maf_path
            self.logger.error(msg)
            raise RuntimeError(msg)
        else:
            self.logger.debug(&#34;Found %d MAF rows from input path %s&#34; % (df.size, maf_path))
        self.metrics = {}
        for index, row in df.iterrows():
            total = row[self.T_DEPTH]
            variants = row[self.T_ALT_COUNT]
            try:
                afp = (float(variants)/float(total))*100
            except ZeroDivisionError:
                afp = &#34;NA&#34;
            oncokb = row.get(self.HIGHEST_LEVEL)
            if oncokb == None:
                oncokb = &#34;NA&#34;
            self.metrics[row[self.HUGO_SYMBOL]] = {
                self.CHROMOSOME: row[self.CHROMOSOME],
                self.PROTEIN_CHANGE: row[self.HGVSP_SHORT],
                self.VARIANT_READS_AND_TOTAL_READS: &#34;%d/%d&#34; % (variants, total),
                self.ALLELE_FRACTION_PERCENTILE: afp,
                self.ONCOKB: oncokb,
                self.FDA_APPROVED_TREATMENT: self._find_treatment_string(row)
            }
        metric_string = json.dumps(self.metrics[row[self.HUGO_SYMBOL]], sort_keys=True)
        self.logger.debug(&#34;Example of metrics for gene %s: %s&#34; % (row[self.HUGO_SYMBOL], metric_string))
        self.logger.info(&#34;Found mutation_extended metrics for %i genes&#34; % len(self.metrics))

    def _find_columns_to_use(self, maf_path):
        &#34;&#34;&#34;
        Read column headers from the MAF file, to determine what columns to input in pandas
        Handles the case where not all OncoKB LEVEL columns are present
        MAF files may have 100+ columns, so this is preferred to reading in the whole file
        &#34;&#34;&#34;
        if re.search(&#34;\.gz$&#34;, maf_path):
            in_file = gzip.open(maf_path, &#39;rt&#39;)
        else:
            in_file = open(maf_path, &#39;r&#39;)
        header_line = None
        # read the first non-comment line to get column headers
        while True:
            line = in_file.readline()
            if line == &#39;&#39;: # end of file
                break
            elif re.match(&#34;#&#34;, line):
                continue
            else:
                header_line = line
                break
        in_file.close()
        if not header_line:
            msg = &#34;No header line found in MAF file &#39;%s&#39;&#34; % maf_path
            self.logger.error(msg)
            raise RuntimeError(msg)
        column_names = re.split(&#34;\t&#34;, line.strip())
        self.logger.debug(&#34;Found MAF column names: &#34;+&#34;, &#34;.join(column_names))
        missing = []
        for name in self.REQUIRED_MAF_COLS:
            if not name in column_names:
                missing.append(name)
        if len(missing) &gt; 0:
            msg = &#34;Missing required MAF columns: &#34;+&#34;, &#34;.join(missing)
            self.logger.error(msg)
            raise RuntimeError(msg)
        use_cols = self.REQUIRED_MAF_COLS.copy() # list of columns to read
        optional_total = 0
        optional_cols = [self.HIGHEST_LEVEL]
        optional_cols.extend(self.BIOMARKER_LEVELS)
        for name in optional_cols:
            if name in column_names:
                use_cols.append(name)
                optional_total += 1
        if optional_total == 0:
            msg = &#34;No optional MAF columns found. OncoKB annotation not applied? &#34;+\
                  &#34;OncoKB output fields will receive &#39;NA&#39; values.&#34;
            self.logger.warning(msg)
        self.logger.debug(&#34;Found MAF columns for input: &#34;+&#34;, &#34;.join(use_cols))
        return use_cols

    def _find_treatment_string(self, row):
        &#34;&#34;&#34;
        Find the &#39;FDA Approved Treatment&#39; string
        Concatenation of level codes and approved treatments
        &#34;&#34;&#34;
        # This string is a temporary placeholder, will be replaced by a more complex structure
        # See https://jira.oicr.on.ca/browse/GCGI-69
        biomarker_values = []
        for level in self.BIOMARKER_LEVELS:
            value = row.get(level) # will be None if no column in MAF file for this level
            if not self.is_null(value):
                biomarker_values.append(&#39;%s:%s&#39; % (level, value))
        if len(biomarker_values) == 0:
            return &#39;NA&#39;
        else:
            return &#39;;&#39;.join(biomarker_values)

    def get_metrics_for_gene(self, gene):
        &#34;&#34;&#34;Get metric dictionary for the named gene, or None if gene is not present&#34;&#34;&#34;
        return self.metrics.get(gene)

    def get_metrics(self):
        return self.metrics</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="djerba.utilities.base.base" href="utilities/base.html#djerba.utilities.base.base">base</a></li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="djerba.metrics.mutation_extended_gene_metrics.ALLELE_FRACTION_PERCENTILE"><code class="name">var <span class="ident">ALLELE_FRACTION_PERCENTILE</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.BIOMARKER_LEVELS"><code class="name">var <span class="ident">BIOMARKER_LEVELS</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.CHROMOSOME"><code class="name">var <span class="ident">CHROMOSOME</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.FDA_APPROVED_TREATMENT"><code class="name">var <span class="ident">FDA_APPROVED_TREATMENT</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.HGVSP_SHORT"><code class="name">var <span class="ident">HGVSP_SHORT</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.HIGHEST_LEVEL"><code class="name">var <span class="ident">HIGHEST_LEVEL</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.HUGO_SYMBOL"><code class="name">var <span class="ident">HUGO_SYMBOL</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_1"><code class="name">var <span class="ident">LEVEL_1</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_2A"><code class="name">var <span class="ident">LEVEL_2A</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_2B"><code class="name">var <span class="ident">LEVEL_2B</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_3A"><code class="name">var <span class="ident">LEVEL_3A</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_3B"><code class="name">var <span class="ident">LEVEL_3B</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_4"><code class="name">var <span class="ident">LEVEL_4</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_R1"><code class="name">var <span class="ident">LEVEL_R1</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.LEVEL_R2"><code class="name">var <span class="ident">LEVEL_R2</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.ONCOKB"><code class="name">var <span class="ident">ONCOKB</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.PROTEIN_CHANGE"><code class="name">var <span class="ident">PROTEIN_CHANGE</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.REQUIRED_MAF_COLS"><code class="name">var <span class="ident">REQUIRED_MAF_COLS</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.T_ALT_COUNT"><code class="name">var <span class="ident">T_ALT_COUNT</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.T_DEPTH"><code class="name">var <span class="ident">T_DEPTH</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.VARIANT_READS_AND_TOTAL_READS"><code class="name">var <span class="ident">VARIANT_READS_AND_TOTAL_READS</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="djerba.metrics.mutation_extended_gene_metrics.get_metrics"><code class="name flex">
<span>def <span class="ident">get_metrics</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_metrics(self):
    return self.metrics</code></pre>
</details>
</dd>
<dt id="djerba.metrics.mutation_extended_gene_metrics.get_metrics_for_gene"><code class="name flex">
<span>def <span class="ident">get_metrics_for_gene</span></span>(<span>self, gene)</span>
</code></dt>
<dd>
<div class="desc"><p>Get metric dictionary for the named gene, or None if gene is not present</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_metrics_for_gene(self, gene):
    &#34;&#34;&#34;Get metric dictionary for the named gene, or None if gene is not present&#34;&#34;&#34;
    return self.metrics.get(gene)</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="djerba.utilities.base.base" href="utilities/base.html#djerba.utilities.base.base">base</a></b></code>:
<ul class="hlist">
<li><code><a title="djerba.utilities.base.base.get_logger" href="utilities/base.html#djerba.utilities.base.base.get_logger">get_logger</a></code></li>
<li><code><a title="djerba.utilities.base.base.is_null" href="utilities/base.html#djerba.utilities.base.base.is_null">is_null</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="djerba.metrics.mutation_extended_sample_metrics"><code class="flex name class">
<span>class <span class="ident">mutation_extended_sample_metrics</span></span>
<span>(</span><span>maf_path, bed_path, tcga_path, cancer_type, log_level=30, log_path=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Sample-level metrics for the MUTATION_EXTENDED alteration type</p>
<p>Includes metrics for Elba output:
- TMB_PER_MB</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class mutation_extended_sample_metrics(base):

    &#34;&#34;&#34;
    Sample-level metrics for the MUTATION_EXTENDED alteration type

    Includes metrics for Elba output:
    - TMB_PER_MB
    &#34;&#34;&#34;
    
    def __init__(self, maf_path, bed_path, tcga_path, cancer_type, log_level=logging.WARNING, log_path=None):
        self.logger = self.get_logger(log_level, &#34;%s.%s&#34; % (__name__, type(self).__name__), log_path)
        self.tmb = self._find_tmb(maf_path, bed_path)        
        [self.tcga_tmb, self.tcga_cohort_tmb] = self._read_tcga(tcga_path, cancer_type)
        self.tcga_pct = ECDF(self.tcga_tmb)(self.tmb)
        self.cohort_pct = ECDF(self.tcga_cohort_tmb)(self.tmb)
        
    def _find_tmb(self, maf_path, bed_path):
        &#34;&#34;&#34;Find the TMB metric: tumor mutation burden / megabase&#34;&#34;&#34;
        names = [&#39;chrom&#39;, &#39;start&#39;, &#39;end&#39;]
        maf = pd.read_csv(maf_path, sep=&#39;\t&#39;, skiprows=1)
        bed = pd.read_csv(bed_path, sep=&#39;\t&#39;, skiprows=2, header=None, names=names)
        target_space = sum(bed[&#39;end&#39;] - bed[&#39;start&#39;]) / 1000000
        keep = [
            &#39;Missense_Mutation&#39;,
            &#39;Frame_Shift_Del&#39;,
            &#39;In_Frame_Del&#39;,
            &#39;Frame_Shift_Ins&#39;,
            &#39;In_Frame_Ins&#39;,
            &#39;Splice_Site&#39;,
            &#39;Translation_Start_Site&#39;,
            &#39;Nonsense_Mutation&#39;,
            &#39;Nonstop_Mutation&#39;
        ]
        tmb = len(maf.loc[maf[&#34;Variant_Classification&#34;].isin(keep)]) / target_space
        return tmb

    def _read_tcga(self, tcga_path, cancer_type):
        &#34;&#34;&#34;Read the TCGA path and return a pair of dictionaries for TMB&#34;&#34;&#34;
        tcga_cols = [&#39;sample&#39;, &#39;tcgaTMB&#39;, &#39;callable&#39;, &#39;cancerType&#39;]
        tcga = pd.read_csv(tcga_path, sep=&#39;\t&#39;, header=0, names=tcga_cols)
        tcga_tmb =  tcga.tcgaTMB
        tcga_cohort = tcga.loc[tcga[&#34;cancerType&#34;] == cancer_type]
        tcga_cohort = tcga_cohort.rename(columns={&#39;tcgaTMB&#39;: &#39;cohortTMB&#39;})
        tcga_cohort_tmb = tcga_cohort.cohortTMB
        return (tcga_tmb, tcga_cohort_tmb)

    def get_cosmic_sigs(self):
        &#34;&#34;&#34;
        TODO: Compute the COSMIC_SIGS metric
        - Currently implemented as an R script
        - Not yet ready for production, requires more validation
        - Return &#34;NA&#34; as a placeholder for now
        &#34;&#34;&#34;
        return &#34;NA&#34;

    def get_tmb(self):
        return self.tmb

    def get_tcga_pct(self):
        return self.tcga_pct
    
    def get_cohort_pct(self):
        return self.cohort_pct
    
    def get_tcga_tmb_as_dict(self):
        &#34;&#34;&#34;Convert the pandas Series to a dictionary&#34;&#34;&#34;
        return self.tcga_tmb.to_dict()

    def get_cohort_tmb_as_dict(self):
        &#34;&#34;&#34;Convert the pandas Series to a dictionary&#34;&#34;&#34;
        return self.tcga_cohort_tmb.to_dict()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="djerba.utilities.base.base" href="utilities/base.html#djerba.utilities.base.base">base</a></li>
</ul>
<h3>Methods</h3>
<dl>
<dt id="djerba.metrics.mutation_extended_sample_metrics.get_cohort_pct"><code class="name flex">
<span>def <span class="ident">get_cohort_pct</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_cohort_pct(self):
    return self.cohort_pct</code></pre>
</details>
</dd>
<dt id="djerba.metrics.mutation_extended_sample_metrics.get_cohort_tmb_as_dict"><code class="name flex">
<span>def <span class="ident">get_cohort_tmb_as_dict</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Convert the pandas Series to a dictionary</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_cohort_tmb_as_dict(self):
    &#34;&#34;&#34;Convert the pandas Series to a dictionary&#34;&#34;&#34;
    return self.tcga_cohort_tmb.to_dict()</code></pre>
</details>
</dd>
<dt id="djerba.metrics.mutation_extended_sample_metrics.get_cosmic_sigs"><code class="name flex">
<span>def <span class="ident">get_cosmic_sigs</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>TODO: Compute the COSMIC_SIGS metric
- Currently implemented as an R script
- Not yet ready for production, requires more validation
- Return "NA" as a placeholder for now</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_cosmic_sigs(self):
    &#34;&#34;&#34;
    TODO: Compute the COSMIC_SIGS metric
    - Currently implemented as an R script
    - Not yet ready for production, requires more validation
    - Return &#34;NA&#34; as a placeholder for now
    &#34;&#34;&#34;
    return &#34;NA&#34;</code></pre>
</details>
</dd>
<dt id="djerba.metrics.mutation_extended_sample_metrics.get_tcga_pct"><code class="name flex">
<span>def <span class="ident">get_tcga_pct</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_tcga_pct(self):
    return self.tcga_pct</code></pre>
</details>
</dd>
<dt id="djerba.metrics.mutation_extended_sample_metrics.get_tcga_tmb_as_dict"><code class="name flex">
<span>def <span class="ident">get_tcga_tmb_as_dict</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Convert the pandas Series to a dictionary</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_tcga_tmb_as_dict(self):
    &#34;&#34;&#34;Convert the pandas Series to a dictionary&#34;&#34;&#34;
    return self.tcga_tmb.to_dict()</code></pre>
</details>
</dd>
<dt id="djerba.metrics.mutation_extended_sample_metrics.get_tmb"><code class="name flex">
<span>def <span class="ident">get_tmb</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_tmb(self):
    return self.tmb</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="djerba.utilities.base.base" href="utilities/base.html#djerba.utilities.base.base">base</a></b></code>:
<ul class="hlist">
<li><code><a title="djerba.utilities.base.base.get_logger" href="utilities/base.html#djerba.utilities.base.base.get_logger">get_logger</a></code></li>
<li><code><a title="djerba.utilities.base.base.is_null" href="utilities/base.html#djerba.utilities.base.base.is_null">is_null</a></code></li>
</ul>
</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="djerba" href="index.html">djerba</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="djerba.metrics.mutation_extended_gene_metrics" href="#djerba.metrics.mutation_extended_gene_metrics">mutation_extended_gene_metrics</a></code></h4>
<ul class="">
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.ALLELE_FRACTION_PERCENTILE" href="#djerba.metrics.mutation_extended_gene_metrics.ALLELE_FRACTION_PERCENTILE">ALLELE_FRACTION_PERCENTILE</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.BIOMARKER_LEVELS" href="#djerba.metrics.mutation_extended_gene_metrics.BIOMARKER_LEVELS">BIOMARKER_LEVELS</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.CHROMOSOME" href="#djerba.metrics.mutation_extended_gene_metrics.CHROMOSOME">CHROMOSOME</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.FDA_APPROVED_TREATMENT" href="#djerba.metrics.mutation_extended_gene_metrics.FDA_APPROVED_TREATMENT">FDA_APPROVED_TREATMENT</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.HGVSP_SHORT" href="#djerba.metrics.mutation_extended_gene_metrics.HGVSP_SHORT">HGVSP_SHORT</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.HIGHEST_LEVEL" href="#djerba.metrics.mutation_extended_gene_metrics.HIGHEST_LEVEL">HIGHEST_LEVEL</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.HUGO_SYMBOL" href="#djerba.metrics.mutation_extended_gene_metrics.HUGO_SYMBOL">HUGO_SYMBOL</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_1" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_1">LEVEL_1</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_2A" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_2A">LEVEL_2A</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_2B" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_2B">LEVEL_2B</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_3A" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_3A">LEVEL_3A</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_3B" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_3B">LEVEL_3B</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_4" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_4">LEVEL_4</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_R1" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_R1">LEVEL_R1</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.LEVEL_R2" href="#djerba.metrics.mutation_extended_gene_metrics.LEVEL_R2">LEVEL_R2</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.ONCOKB" href="#djerba.metrics.mutation_extended_gene_metrics.ONCOKB">ONCOKB</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.PROTEIN_CHANGE" href="#djerba.metrics.mutation_extended_gene_metrics.PROTEIN_CHANGE">PROTEIN_CHANGE</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.REQUIRED_MAF_COLS" href="#djerba.metrics.mutation_extended_gene_metrics.REQUIRED_MAF_COLS">REQUIRED_MAF_COLS</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.T_ALT_COUNT" href="#djerba.metrics.mutation_extended_gene_metrics.T_ALT_COUNT">T_ALT_COUNT</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.T_DEPTH" href="#djerba.metrics.mutation_extended_gene_metrics.T_DEPTH">T_DEPTH</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.VARIANT_READS_AND_TOTAL_READS" href="#djerba.metrics.mutation_extended_gene_metrics.VARIANT_READS_AND_TOTAL_READS">VARIANT_READS_AND_TOTAL_READS</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.get_metrics" href="#djerba.metrics.mutation_extended_gene_metrics.get_metrics">get_metrics</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_gene_metrics.get_metrics_for_gene" href="#djerba.metrics.mutation_extended_gene_metrics.get_metrics_for_gene">get_metrics_for_gene</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="djerba.metrics.mutation_extended_sample_metrics" href="#djerba.metrics.mutation_extended_sample_metrics">mutation_extended_sample_metrics</a></code></h4>
<ul class="">
<li><code><a title="djerba.metrics.mutation_extended_sample_metrics.get_cohort_pct" href="#djerba.metrics.mutation_extended_sample_metrics.get_cohort_pct">get_cohort_pct</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_sample_metrics.get_cohort_tmb_as_dict" href="#djerba.metrics.mutation_extended_sample_metrics.get_cohort_tmb_as_dict">get_cohort_tmb_as_dict</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_sample_metrics.get_cosmic_sigs" href="#djerba.metrics.mutation_extended_sample_metrics.get_cosmic_sigs">get_cosmic_sigs</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_sample_metrics.get_tcga_pct" href="#djerba.metrics.mutation_extended_sample_metrics.get_tcga_pct">get_tcga_pct</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_sample_metrics.get_tcga_tmb_as_dict" href="#djerba.metrics.mutation_extended_sample_metrics.get_tcga_tmb_as_dict">get_tcga_tmb_as_dict</a></code></li>
<li><code><a title="djerba.metrics.mutation_extended_sample_metrics.get_tmb" href="#djerba.metrics.mutation_extended_sample_metrics.get_tmb">get_tmb</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>