
<p> This assay combines two comprehensive next generation sequencing assays: a targeted sequencing (TAR) assay and a shallow whole genome sequencing (sWGS) assay.</p>
      
<p>
     TS libraries are prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE, cfDNA, fresh frozen tissue (for tumour samples) or buffy coat blood specimens (for matched normal blood samples). 
     Paired-end sequencing is performed using the NextSeq 550 technology. 
     For targeted sequencing, alignments are performed using <a href=https://bio-bwa.sourceforge.net/>bwa</a> mem (0.7.12) against reference genome <a href=https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.38/>GRCh38.p12</a>. 
     Unique molecule based error suppression is completed with ConsensusCruncher (https://github.com/pughlab/ConsensusCruncher). 
     SNVs and INDELs are called using MuTect2 (GATK 4.1.1.0) and annotated with VariantEffectPredictor (105.0) followed by OncoKB. 
     In addition, variants without an OncoKB actionable tier are reported if they are listed as oncogenic, likely oncogenic or predicted oncogenic. 
</p>   
  
<p>
    Shallow whole genome sequencing libraries were prepared using the KAPA Hyper Prep kit with DNA extracted from FFPE, cfDNA, fresh frozen tissue (for tumour samples). 
    Paired-end sequencing is performed using the NextSeq 550 technology to a minimum target coverage of 0.1x. 
    Alignments are performed using <a href=https://bio-bwa.sourceforge.net/>bwa</a> mem (0.7.12) against reference genome <a href=https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.38/>GRCh38.p12</a> 
    and copy number amplifications are called using ichorCNA <a href=https://github.com/broadinstitute/ichorCNA>ichorCNA</a>. 
</p>


<p>
    Assay results were collated into the report document by <a href=constants.DJERBA_LINK>Djerba</a> (djerba v.${constants.DJERBA_VERSION}, pWGS plugin v.${constants.PWGS_DJERBA_VERSION}) 
    using pipeline ${constants.DJERBA_PIPELINE_VERSION}.
  </p>
