<!--This file is the HTML code for the sWGS section of the report.-->
<!--AUTHOR: Aqsa Alam-->

<!-- IMPORTS -->
<%
  from djerba.plugins.tar.swgs.render import html_builder
  import djerba.plugins.tar.swgs.constants as constants
  from djerba.util.image_to_base64 import converter
%>
  
  
<!-- Copy Number Variants -->
${html_builder().section_cells_begin("<h2>Shallow Whole Genome Sequencing</h2>","main")}


       % if results.get(constants.PASS_TAR_PURITY):
          <p>      
                The alterations below are oncogenic as defined by OncoKB. 
                Small regions (&#60;3mb) with large copy number gains marked as &#9650 in plot below.
          </p>
       % else:
          <p>
                As the inferred tumour purity is &lt;10&#37;, copy number amplifications cannot be inferred. 
                Small regions (&#60;3mb) with large copy number gains marked as &#9650 in plot below.
          </p>


       % endif

       <img id='sWGS' style='width: 100%; object-fit: contain' src="${converter().convert_svg(results.get(constants.CNV_PLOT), 'CNV plot')}"/>

      % if results.get(constants.PASS_TAR_PURITY):
          % if results.get(constants.CLINICALLY_RELEVANT_VARIANTS) > 0:
                <table class="variants" width="100%">
                ${html_builder().oncogenic_CNVs_header(results)}
                <tbody>
                % for row in html_builder().oncogenic_CNVs_rows(results):
                      ${row}
                % endfor
                </table>
                </tbody>
                </table>
	        </table>

          % endif
      % endif
      
${html_builder().section_cells_end()}
