<%
  from time import strftime
  from djerba.plugins.pwgs.analysis.render.json_to_html import html_builder
  import djerba.plugins.pwgs.constants as constants
%>



<!-- ctDNA Detection -->
${html_builder().section_cells_begin("<h2>Mutational Sampling</h2>","main")}

      <img id='Detection' style='width: 100%; object-fit: contain' src="${results.get(constants.PWGS_PLOT)}"/>

      <p> This assay counted the number of sites in the patient's plasma 
            at which variants were detected that match variants found in the patient's primary tumour 
            (candidate sites; <strong>${html_builder().k_comma_format(results.get(constants.SITES_CHECKED))}</strong>).
            However, candidate SNVs occasionally match SNVs arising from sequencing artefacts. 
            To test for this possibility, the number of detected sites in the plasma sample is statistically 
            compared to the number of detected sites in a healthy Control cohort. 
            The number of SNVs of tumour origin 
            detected in this plasma sample (<strong>${html_builder().k_comma_format(results.get(constants.SITES_DETECTED))}</strong>) was
             <strong>${results.get(constants.SIGNIFICANCE)} </strong>  
             than in the control cohort (N=${results.get(constants.COHORT_N)}).</p>
            <table class="variants" style="width:100%">
                  <thead>
                        <th style=" width:16%">Reads Checked</th>
                        <th style=" width:16%">Reads Detected</th>
                        <th style=" width:20%">Detected / Checked (%)</th>
                  </thead>
                  <tbody>
                        <tr>
                              <td>${html_builder().k_comma_format(results.get(constants.READS_CHECKED))}</td>
                              <td>${html_builder().k_comma_format(results.get(constants.READS_DETECTED))}</td>
                              <td>${results.get(constants.TUMOUR_FRACTION_READS)}</td>
                        </tr>
                  </tbody>
            </table>

${html_builder().section_cells_end()}