
<hr class="header-big-line" style="width: 100%; align="right" >
  <h2>Supplementary </h2>



  <hr class="big-white-line" ><div class="twocellsupp"><div class="oneoftwocellsupp"><h3>Definitions</h3></div><div class="twooftwocellsupp" ><hr class="big-line" >

<p>
  <strong>Coverage: </strong>Mean number of bases covering each sequenced base. Minimum coverage threshold is 30x (with a target of 40x).
  <br><strong>Healthy Blood Controls (HBCs):</strong> Something about them, how they were sequenced etc.
</p>

  </div></div>

  <hr class="big-white-line" ><div class="twocellsupp"><div class="oneoftwocellsupp"><h3>Assay description</h3></div><div class="twooftwocellsupp" ><hr class="big-line" >

    <p>This assay is a comprehensive next generation sequencing assay, specifically plasma whole genome sequencing (WGS). 
        Plasma libraries are prepared using the KAPA Hyper Prep kit with DNA extracted from cfDNA. 
        Paired-end sequencing is performed using the NovaSeq6000 technology. 
        Alignments are performed using <a href="https://bio-bwa.sourceforge.net/">bwa</a> mem (0.7.12) against reference genome <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39/">GRCh38.p13</a> and processed with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-">Picard</a> (2.21.2), which includes duplicate marking. 
        Final detection rate is produced by <a href="https://pubmed.ncbi.nlm.nih.gov/32483360/">MRDetect</a> (1.0).
    Assay results were collated into the report document by <a href="https://github.com/oicr-gsi/djerba">Djerba</a> (0.3.20) using pipeline v.1.0.</p>
  </div></div>

  <hr class="big-white-line" ><div class="twocellsupp"><div class="oneoftwocellsupp"><h3>Disclaimer</h3></div><div class="twooftwocellsupp" ><hr class="big-line" >
  <p>Based on a minimum coverage of 30, the sensitivity is XX%. The limit of detection is XX% VAF for SNVs.</p>
  <p><strong>Genes tested</strong>: Plasma whole genome sequencing encompasses all genes and intergenic regions in a specimen, and gene annotation is not performed in this assay.</p>

<p>This test was developed and its performance characteristics determined by OICR Genomics. It has not been cleared or approved by the US Food and Drug Administration.</p>

</div></div>
  
  <hr class="lil-line" style="width: 100%;"  >
  <h3>Report Sign offs</h3>

  <table class="suppl" width="100%">
    <tr>
      <td width="33%">Report drafted by Felix Beaudry on 2023/02/07</td>
    </tr>
    <tr>
      <td width="33%">Report electronically signed out by Trevor Pugh, PhD, FACMG (ABMS #1027812) on yyyy/mm/dd</td>
    </tr>
  </table>

  <br>
  <br>
  <br>
  <br>

  <hr class="big-line" style="width: 100%; "  >
