<%
  from time import strftime
  from djerba.plugins.pwgs.analysis.render.json_to_html import html_builder
  import djerba.plugins.pwgs.constants as constants
%>

<hr class="header-big-line" style="width: 100%; ; margin: 400px 0px 0px 0px;" >
<h2>Supplementary </h2>

${html_builder().section_cells_begin("<h3>Definitions</h3>","supp")}

<p>
  <br><strong>Minimal Residual Disease (MRD): </strong>malignant cells that remain in a patient who achieves remission
  and that share genetics with the original tumour cells (<em>see</em> DOI: <a href="https://pubmed.ncbi.nlm.nih.gov/29376520/">10.1038/nrc.2017.125</a>). 
  Detection of minimal residual disease is not conclusively associated with prognosis, 
  treatment choices should be made on a case-by-case basis with the treating physician.

  <br><strong>circulating tumour DNA (ctDNA): </strong>DNA that comes from dying cancerous cells and is found in the bloodstream. Here, a marker for MRD. 

  <br><strong>Coverage: </strong>the mean number of bases covering each sequenced base. Minimum coverage threshold is 30x (with a target of 40x).

  <br><strong>Insert Size: </strong>the length in base pairs of the DNA between adapters, affected by both biology and laboratory. 

  <br><strong>Primary SNVs: </strong>the number of candidate single nucleotide variants from the primary tumour. 
  If primary SNVs is too low (below 4,000), mutatoinal sampling analysis is unreliable.

  <br><strong>Mutational Sampling Analysis: </strong>sampling and sequencing cell-free DNA in plasma 
  to search for single nucleotide variants matching the primary tumour. 
  
  <br><strong>Detected-Checked Reads Ratio: </strong>the percentage of reads with primary tumour matches divided by the total checked reads covering candidate sites.
  When ctDNA is detected, it may be an estimate of the fraction of ctDNA in the blood, but if ctDNA is not detected it is an estimate of sequencing artefact matches.
</p>

${html_builder().section_cells_end()}

${html_builder().section_cells_begin("<h3>Assay description</h3>","supp")}

<p>Plasma whole genome sequencing (pWGS) is a comprehensive next generation sequencing assay for the detection of minimal residual disease. 
  Plasma libraries were prepared using the KAPA Hyper Prep kit with DNA extracted from cfDNA. 
  Paired-end sequencing was performed using the NovaSeq6000 technology. 
  Alignments were performed using <a href=${constants.BWAMEM_LINK}>bwa</a> mem (${constants.BWAMEM_VERSION}) 
  against reference genome <a href=${constants.REFERENCE_GENOME_LINK}>${constants.REFERENCE_GENOME_VERSION}</a> 
  and processed with <a href=${constants.PICARD_LINK}>Picard</a> (${constants.PICARD_VERSION}), which includes duplicate marking. 
  Reads were filtered for ctDNA likelihood and matches with primary tumour by <a href=${constants.MRDETECT_LINK}>MRDetect</a> (${constants.MRDETECT_VERSION}). 
  The sample is compared against a control cohort of Healthy Blood Controls (HBCs) sequenced and processed at OICR using the same procedure as any tested plasma sample.
  Assay results were collated into the report document by <a href=constants.DJERBA_LINK>Djerba</a> (${constants.DJERBA_VERSION}) using pipeline ${constants.DJERBA_PIPELINE_VERSION}.
</p>

<br>
<p></p>

${html_builder().section_cells_end()}

${html_builder().section_cells_begin("<h3>Disclaimer</h3>","supp")}
<p>
Based on a minimum coverage of 30X, the sensitivity is 93%. The limit of detection is 0.04% ctDNA.
<br>Minimal Residual Disease testing by Plasma whole genome sequencing encompasses all genomic regions in a specimen. 
Gene annotation is not performed in this assay and variants are not classified (e.g. by pathogeneity). 
<br>This test was developed and its performance characteristics determined by OICR Genomics. 
It has not been cleared or approved by the US Food and Drug Administration.
</p>

${html_builder().section_cells_end()}
  
<hr class="lil-line" style="width: 100%;"  >
<h3>Report Sign offs</h3>

  <table class="suppl" width="100%">
    <tr>
      <td width="33%">Report drafted by ${results.get(constants.AUTHOR)} on ${strftime('%Y/%m/%d')}</td>
    </tr>
    <tr>
      <td width="33%">Report electronically signed out by Trevor Pugh, PhD, FACMG (ABMS #1027812) on yyyy/mm/dd</td>
    </tr>
  </table>

  <br>
  <br>
  <br>
  <br>

  <hr class="big-line" style="width: 100%; "  >
