---
output:
  html_document:
    includes:
      before_body: header.html
      after_body: !expr Sys.getenv('DJERBA_RMD_AFTER_BODY')
    number sections: yes
---

```{r, echo = FALSE, message= FALSE, warning=FALSE}
#Vectorize(require)(package = c("knitr", "tidyr", "kableExtra", "stringr", "readr",
#"dplyr", "tibble", "ggplot2"), char = TRUE)

library(knitr)
library(tidyr)
library(kableExtra)
library(stringr)
library(readr)
library(dplyr)
library(tibble)
library(ggplot2)
library(data.table)

args <- commandArgs(trailingOnly = T)

#folder <- '/.mounts/labs/TGL/cap/OCTCAP/OCT_011408/OCT_011408_Co_P_OCT_011408_TS/report'
#folder <- '/.mounts/labs/TGL/cap/OCTCAP/OCT_010118/OCT_010118_Ut_P_OCT-01-0118-TS/report'

#cgiTools <- "/u/afortuna/CGI-Tools"
cgiTools <- Sys.getenv("DJERBA_BASE_DIR")

folder <- args[1]

# Patient Summary
#patSum <- readLines(paste(folder, "genomic_summary.txt", sep = "/"))
qcStat <- readLines(paste(folder, "genomic_summary.txt", sep = "/"))
# annoations
civicDataDir <- paste(cgiTools, "data/civic", sep = "/")

# OncoKB order
target <- c("Level 1", "Level 2", "Level 3A", "Level 3B", "Level 4", "Level R1",
  "Level R2", "Oncogenic", "Likely Oncogenic")

levels <- c("LEVEL_1", "LEVEL_2", "LEVEL_3A", "LEVEL_3B", "LEVEL_4", "LEVEL_R1",
  "LEVEL_R2", "LEVEL_R3")


# clinical data should be parsed from data_clinical_patients
clinicalData <- read.csv(paste(folder, "data_clinical.txt", sep = "/"), stringsAsFactors = F, sep = "\t")

# sampleTMB <-
# round(mutationCount/((clinicalData$PCT_V7_ABOVE_80X/100)*37.285536),2)
sample_tcga <- tolower(ifelse(clinicalData$CLOSEST_TCGA == "COAD" | clinicalData$CLOSEST_TCGA ==
  "READ", "coadread", clinicalData$CLOSEST_TCGA))

```

<style>
#example1 {
  background: #65bc45 !important;
  border-radius: 10px !important;
  padding-left: 10px !important;
  padding-bottom: 10px !important;
  color: white !important;
  font-size: 30px !important;
  font-weight: bold !important;
  height: 30px !important;
  line-height: 30px !important;
  font-family: Arial !important;
  display: block;
  clear: both
}

#example2 {
  background: #65bc45 !important;
  padding-left: 10px !important;
  padding-bottom: 10px !important;
  color: white !important;
  font-size: 30px !important;
  font-weight: bold !important;
  height: 4px !important;
  line-height: 10px !important;
  font-family: Arial !important;
  display: block;
  clear: both
}

#example3 {
  padding-left: 10px !important;
  padding-bottom: 10px !important;
  color: black !important;
  font-size: 65px !important;
  font-weight: bold !important;
  text-align: center;
  font-family: Arial !important;
  display: block;
  clear: both
}

.clear {
  display: block;
}

h3.clear {
  clear: both;
}

html * {
  font-family: Arial !important;
}

* {
  box-sizing: border-box;
}

.column {
  float: left;
  padding-left: 25px;
  padding-right: 25px;
  width: 33.33%;
}

.column-2 {
  float: right;
  style=width: 200px;
  padding: 0px 0px 0px 20px;
}

.row:after {
  content: "";
  display: table;
  clear: both;
}

@media print {
  .pagebreak {
  page-break-before: always;
 }
}

p {
  margin: 0.25em;
}

section {
  display: flex;
  padding-top: 0.5rem;
}

.bold {
  font-weight: 600;
}

@page {
  size: 8.5in 11in;
}

.bold {
  font-weight: 600;
}

td, th {
  text-align: left;
  padding: 2px;
}

</style>

<html>

<div id="example3">
  <h3>QUALITY FAILURE REPORT</h3>
</div>

<div id="example2">
  <h3></h3>
</div>

<table width="100%">
  <tr>
    <td><strong>Assay: </strong>Tumour-normal whole genome and tumour transcriptome sequencing (WGTS, v1.0)</td>
  </tr>
</table>
<table width="100%">
  <tr>
    <td width="33%"><strong>Date of Report:</strong> `r format(Sys.time(), '%d %B, %Y')`</td>
    <td width="33%"><strong>Requisition Approved:</strong> `r format(Sys.time(), '%d %B, %Y')`</td>
    <td width="33%"><strong>Report Identifier:</strong> `r paste0(clinicalData$SAMPLE_ID, " (v", clinicalData$REPORT_VERSION, ")")`</td>
  </tr>
  <tr>
    <td width="33%"><strong>Study:</strong> PASS-01</td>
    <td width="33%"><strong>Patient Study ID:</strong> `r clinicalData$SAMPLE_ID`</td>
    <td width="33%"><strong>Assay ID:</strong> `r clinicalData$SAMPLE_ID`</td>
  </tr>
  <tr>
    <td width="33%"><strong>Name:</strong> LAST NAME, FIRST NAME</td>
    <td width="33%"><strong>DOB:</strong> yyyy/mm/dd</td>
    <td width="33%"><strong>Genetic Sex:</strong> `r clinicalData$SEX`</td>
  </tr>
  <tr>
    <td width="33%"><strong>Physician:</strong> LAST, FIRST</td>
    <td width="33%"><strong>Licence #:</strong> nnnnnnnn</td>
    <td width="33%"><strong>Phone #:</strong> nnn-nnn-nnnn</td>
  </tr>
</table>
<table width="100%">
  <tr>
    <td width="50%"><strong>Hospital:</strong> HOSPTIAL NAME AND HOSPITAL ADDRESS</td>
    <td width="50%"><strong>Requisitioner Email:</strong> NAME@DOMAIN.COM</td>
  </tr>
  <tr>
    <td width="50%"><strong>Primary cancer:</strong> `r clinicalData$CANCER_TYPE_DESCRIPTION`</td>
    <td width="50%"><strong>Site of biopsy/surgery:</strong> `r clinicalData$SAMPLE_ANATOMICAL_SITE`</td>
  </tr>
  <tr>
    <td width="50%"><strong>Tumour Sample ID:</strong> nnnnnnnn</td>
    <td width="50%"><strong>Blood Sample ID:</strong> nnnnnnnn</td>
  </tr>
</table>

<div id="example1">
  <h3>Sample Information & Quality</h3>
</div>

<table width="100%">
  <tr>
    <td width="33%"><strong>OncoTree code:</strong> `r clinicalData$CANCER_TYPE`</td>
    <td width="33%"><strong>Callability (%):</strong> `r clinicalData$PCT_V7_ABOVE_80X`</td>
    <td width="33%"><strong>Coverage (mean):</strong> `r clinicalData$MEAN_COVERAGE`</td>
  </tr>
  <tr>
    <td width="33%"><strong>Sample Type:</strong> `r clinicalData$SAMPLE_TYPE`</td>
    <td width="33%"><strong>Estimated Cancer Cell Content (%):</strong> `r clinicalData$SEQUENZA_PURITY_FRACTION * 100`</td>
    <td width="33%"><strong>Estimated Ploidy:</strong> `r clinicalData$SEQUENZA_PLOIDY`</td>
  </tr>
</table>


<div id="example1">
  <h3>Summary of Quality Control Findings</h3>
</div>
<br>
<p style="border:3px; border-style:solid; border-color:red; padding: 1em;">`r qcStat`</p>

</html>

<div id="example2">
  <h3></h3>
</div>
